Update Dependencies + Set Java compiler version to 11 (requirement OpenSaml4) + add methods for Jakarta HttpServlet

csonnebo opened the pull request:
Includes three implementations:
* Update all dependencies --> No impact on existing code
* Set Java compiler version to 11 --> OpenSaml 4 requires java 11 anyway
* Clone serveral Methode for use with **javax**.servlet.http.HttpServletRequest or **jakarta**.servlet.http.HttpServletRequest --> For supporting Jakarta Servlet 5.0

justinbleach commented:
@csonnebo do you know when these changes might be merged in? I'm upgrading from Springboot 2 to Springboot 3 and all the dependencies in SB 3 use Jarkata instead of javax.

csonnebo commented:
Unfortunately I have no Idea. I am planing to upgrade to spring 6 as well,
so I am waiting for my changes too.

Is the a way to contact the owner of the project and tell that we need
theses changes?

An update of all dependencies to reduce vulnerabilities would be again a
good idea.


Am 21. September 2023 19:32:25 schrieb Justin Bleach
***@***.***>:
> @csonnebo do you know when these changes might be merged in? I'm upgrading
> from Springboot 2 to Springboot 3 and all the dependencies in SB 3 use
> Jarkata instead of javax.
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.

justinbleach commented:
@JPLachance @dreisch-coveo any chance of a review and merge to update this library with the latest dependencies. I personally really need the Jakarta dependencies update for this to work with Springboot 3

jebeaudet commented on src/main/java/com/coveo/saml/ValidatorUtils.java:
```diff
@@ -16,12 +17,19 @@
  */
 class ValidatorUtils {
 
+  private static final Logger logger = LoggerFactory.getLogger(SamlClient.class);
```
```suggestion
  private static final Logger logger = LoggerFactory.getLogger(ValidatorUtils.class);
```

csonnebo replied:
I would prefer to leave the logger on com.coveo.saml.SamlClient and not open a new logger. I would not split the log entries further but keep them together. 
What do you think?

jebeaudet commented on src/main/java/com/coveo/saml/ValidatorUtils.java:
```diff
@@ -34,6 +42,7 @@ private static void validateResponse(StatusResponseType response, String respons
     try {
       new ResponseSchemaValidator().validate(response);
     } catch (SamlException e) {
+      logger.debug("The response schema validation failed", e);
```
I know this is debug but this usually means this will be double logged since the throwable is already wrapped in another. Any specific reason why you added this?

csonnebo replied:
Good point. I think we can remove that.

jebeaudet commented on pom.xml:
```diff
@@ -54,18 +54,25 @@
+      <groupId>jakarta.servlet</groupId>
+      <artifactId>jakarta.servlet-api</artifactId>
+      <version>6.0.0</version>
+      <scope>provided</scope><!-- Must be specified as provided because of the license -->
```
I don't see the licensing issue, especially with jakarta, care to elaborate?

csonnebo replied:
The Jakarta Servlet API is licensed under EPL 2.0 which is critical and under GPL which is realy strong critical . If you use OSS under GPL license your prooct have to be licensed under GPL as well and this is very bad. That is why I recommend to iclude it just with scope provided.

I would also advise putting this Saml client library under Apache 2.0 licence. The Apache 2.0 licence is the least problematic.
But that's your topic ;-)

jebeaudet replied:
MIT license is definitely the most permissive though.

IANAL so I've reached out to our legal team for regarding your other comment.

justinbleach replied:
@ jebeaudet any movement on this from legal?

jebeaudet replied:
I got the feedback from our lawyers that including jakarta.servlet-api is fine in the current circumstances. 

I'll get things moving for this PR internally, but I cannot make any promise nor ETA.

justinbleach replied:
Thanks for working to get this done. Love the library so glad it will keep up with Java 17 changes!

jebeaudet commented on pom.xml:
```diff
@@ -54,18 +54,25 @@
 
   <dependencies>
     <dependency>
       <groupId>javax.servlet</groupId>
```
@malaporte I really think this should be a breaking release and get rid of the javax stuff entirely

jebeaudet commented on src/main/java/com/coveo/saml/ValidatorUtils.java:
```diff
@@ -248,6 +257,11 @@ public static void validate(
     validateSignature(response, credentials);
   }
 
+  public static void validateIdpMetadata(InputStream metadata) throws SamlException, XMLParserException {
```
this does not seem used, which mean the `createDOMParser` could be moved back to its original location.

csonnebo replied:
I only wanted to create a possibility to validate Idp metadata in the library and to maintain this possibility here instead of implementing it externally myself.

If you want to remove it, it's okay for me.

jebeaudet replied:
At this point, I'd like to keep the public api as it was so yes, please remove it.

csonnebo commented:
I commented on @jebeaudet's comments. What happens now?

csonnebo commented:
@jebeaudet: I made the corrections as discussed. Please come back to me with the results regarding the jakarta.servlet-api licence and if we should remove all the javax stuff. 
I also updated the dependencies once more and added a maven plugin for that. Please review that changes as well.

Thank you very much.

jebeaudet commented:
We understand that you want this merged and released but I'm asking you some more patience while we figure some things out internally about this library, on top of the legal issue that is still under evaluation. Thanks!

justinbleach commented:
@jebeaudet hope things are going well. Here is an example of another repo that uses Jarkata with an apache 2.0 license:
https://github.com/mkopylec/charon-spring-boot-starter maybe that will help with your license discussions

csonnebo commented:
@jebeaudet I set servlet-api scope to compile.
What do you think about removing the javax.servlet-api?

csonnebo commented:
@jebeaudet Any news here?
I think we should remove javax.servlet-api. I think otherwise you have to provide both libs (javax and jarkarta) in production. If one is missing you will get a compiler error.

What do you think?

csonnebo commented:
@jebeaudet 
Hi again,
Obviously the dependency review check fails. What steps are necessary next?

jebeaudet commented:
Hi @csonnebo 

Thanks for the ping. I've merged the PR with one major change (javax.servlet is completely dropped) and some minor mainly cosmetic changes.

I'll be releasing this as a major release shortly.

Thanks again for your contribution and sorry for the back and forth on this, this library has no clear owner internally and this will get addressed.
